File: csharp/queen-attack/QueenAttack.cs
using System;

public class Queen
{
    public Queen(int row, int column)
    {
        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }
}

public static class Queens
{
    public static bool CanAttack(Queen white, Queen black)
    {
        if(white.Row == black.Row && white.Column == black.Column)
        {
            throw new ArgumentException($"The spot {white.Column}, {white.Row} is already occupied.");
        }

        return white.Row == black.Row || white.Column == black.Column || Math.Abs(white.Column - black.Column) == Math.Abs(white.Row - black.Row);
    }

}

File: lua/queen-attack/queen-attack.lua
return function(pos)
    assert(-1 < pos.row and pos.row < 8)
    assert(-1 < pos.column and pos.column < 8)

    return {
        pos = pos,
        can_attack = function(other)
            return pos.row == other.pos.row or pos.column == other.pos.column or
            math.abs(pos.column - other.pos.column) == math.abs(pos.row - other.pos.row)
        end
    }
end